torch/_functorch/aot_autograd.py

```diff
@@ -273,343 +552,166 @@ def from_fun(t):
-            inp_tensor_ids = {id(inpt) for inpt in flat_f_args if isinstance(inpt, torch.Tensor)}
-            inp_storage_refs_set = set(inp_storage_refs)
-
-            non_aliased_input_outs = []
-            # For a given output tensor that alias an input, tells us:
-            # (1) the index of the input that we alias
-            # (2) Whether or not the output is a view of the input, or if `output is input`
-            #     (so we don't need to generate a view, and can return the input directly)
-            # Note: if the function returns an output that *is* an input, we still cannot return it in the graph.
-            # e.g.
-            #   def f(x):
-            #       x.add_(1)
-            #       return x
-            # Our compiled fw will return an "x_updated", but it is *not* ok to return that to the user.
-            # We need to manually do x.copy_(x_updated), and return the original x to the user.
-            # Why? for example, the metadata between x and x_updated might be different (e.g. _is_leaf())
-            aliased_out_idx: Dict[torch.Tensor, Tuple[int, bool]] = {}
-
-            for o in outputs:
-                # Note: When detecting input/output aliasing, we NEED to do it using the outer FunctionalTensorWrapper objects.
-                # In the case where we mutate an input *and* return a view of it, the outer wrappers will still alias,
-                # but the inner tensors no longer alias.
-                if isinstance(o, torch.Tensor) and StorageWeakRef(o._storage()) in inp_storage_refs:
-                    aliased_inp_idx = inp_storage_refs[StorageWeakRef(o._storage())]
-                    is_exact_input = id(o) in inp_tensor_ids
-                    aliases_intermediate_and_not_input = False
-                    aliased_out_idx[o] = (
-                        aliased_inp_idx,
-                        aliases_intermediate_and_not_input,
-                        is_exact_input,
-                    )
+        num_mutated_inps = len(
+            [x for x in input_info if x.mutates_data or x.mutates_metadata]
+        )
+        inp_storage_refs = {
+            StorageWeakRef(inpt._storage()): idx
+            for idx, inpt in enumerate(flat_f_args)
+            if isinstance(inpt, torch.Tensor)
+        }
+
+        # We need inp tensor id's to be able to tell if an outputs **are** inputs.
+        inp_tensor_ids = {
+            id(inpt) for inpt in flat_f_args if isinstance(inpt, torch.Tensor)
+        }
+        # We need output tensor id's to tell if any output._base` attributes **are** other outputs.
+        # (This is also a dict because we need to know that output's index, so we can regenerate
+        # the alias from it).
+        out_tensor_ids = {id(o): i for i, o in enumerate(flat_f_outs)}
+        # maps the id of an intermediate base to its index in the output of the compiled forward
+        intermediate_base_tensor_id_to_output_idx: Dict[int, int] = {}
```

Comment:
NewTypes would be really handy here, the semantic mismatch of the int, int here is very jarring